man/aria_dump_log.1
.TH ARIA_DUMP_LOG "1" "May 2014" "aria_dump_log Ver 1.0" "User Commands"
.SH NAME
aria_dump_log \- Dump content of Aria log pages.
.SH SYNOPSIS
.B aria_dump_log
\fI-f file OPTIONS\fR
.SH DESCRIPTION
Dump content of Aria log pages.
.TP
\fB\-#\fR, \fB\-\-debug\fR[=\fIname\fR]
Output debug log. Often the argument is 'd:t:o,filename'.
.TP
\fB\-f\fR, \fB\-\-file\fR=\fIname\fR
Path to file which will be read
.TP
\fB\-?\fR, \fB\-\-help\fR
Display this help and exit.
.TP
\fB\-o\fR, \fB\-\-offset=\fR#
Start reading log from this offset
.TP
\fB\-n\fR, \fB\-\-pages=\fR#
Number of pages to read
.TP
\fB\-U\fR, \fB\-\-unit\-test\fR
Use unit test record table (for logs created by unittests
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit.
.PP
Default options are read from the following files in the given order:
\fB/etc/my.cnf\fB
\fB/etc/mariadb/my.cnf\fB
\fB~/.my.cnf\fB
.PP
The following groups are read: \fBaria_dump_log\fR
.PP
The following options may be given as the first argument:
.TP
\fB\-\-print\-defaults\fR
Print the program argument list and exit.
.TP
\fB\-\-no\-defaults\fR
Don't read default options from any option file.
.TP
\fB\-\-defaults\-file=\fR#
Only read default options from the given file #.
.TP
\fB\-\-defaults\-extra\-file=\fR#
Read this file after the global files are read.
.PP
.SH "SEE ALSO"
For more information, please refer to the MariaDB Documentation, available online at https://mariadb.com/docs/
.SH AUTHOR
MariaDB Foundation (http://www.mariadb.org/).


pcre/NEWS
News about PCRE releases
------------------------

Note that this library (now called PCRE1) is now being maintained for bug fixes
only. New projects are advised to use the new PCRE2 libraries.


Release 8.43 23-February-2019
-----------------------------

This is a bug-fix release.


Release 8.42 20-March-2018
--------------------------

This is a bug-fix release.


Release 8.41 13-June-2017
-------------------------

This is a bug-fix release.


Release 8.40 11-January-2017
----------------------------

This is a bug-fix release.


Release 8.39 14-June-2016
-------------------------

Some appropriate PCRE2 JIT improvements have been retro-fitted to PCRE1. Apart
from that, this is another bug-fix release. Note that this library (now called
PCRE1) is now being maintained for bug fixes only. New projects are advised to
use the new PCRE2 libraries.


Release 8.38 23-November-2015
-----------------------------

This is bug-fix release. Note that this library (now called PCRE1) is now being
maintained for bug fixes only. New projects are advised to use the new PCRE2
libraries.


Release 8.37 28-April-2015
--------------------------

This is bug-fix release. Note that this library (now called PCRE1) is now being
maintained for bug fixes only. New projects are advised to use the new PCRE2
libraries.


Release 8.36 26-September-2014
------------------------------

This is primarily a bug-fix release. However, in addition, the Unicode data
tables have been updated to Unicode 7.0.0.


Release 8.35 04-April-2014
--------------------------

There have been performance improvements for classes containing non-ASCII
characters and the "auto-possessification" feature has been extended. Other
minor improvements have been implemented and bugs fixed. There is a new callout
feature to enable applications to do detailed stack checks at compile time, to
avoid running out of stack for deeply nested parentheses. The JIT compiler has
been extended with experimental support for ARM-64, MIPS-64, and PPC-LE.


Release 8.34 15-December-2013
-----------------------------

As well as fixing the inevitable bugs, performance has been improved by
refactoring and extending the amount of "auto-possessification" that PCRE does.
Other notable changes:

.  Implemented PCRE_INFO_MATCH_EMPTY, which yields 1 if the pattern can match
   an empty string. If it can, pcretest shows this in its information output.

.  A back reference to a named subpattern when there is more than one of the
   same name now checks them in the order in which they appear in the pattern.
   The first one that is set is used for the reference. Previously only the
   first one was inspected. This change makes PCRE more compatible with Perl.

.  Unicode character properties were updated from Unicode 6.3.0.

.  The character VT has been added to the set of characters that match \s and
   are generally treated as white space, following this same change in Perl
   5.18. There is now no difference between "Perl space" and "POSIX space".

.  Perl has changed its handling of \8 and \9. If there is no previously
   encountered capturing group of those numbers, they are treated as the
   literal characters 8 and 9 instead of a binary zero followed by the
   literals. PCRE now does the same.

.  Following Perl, added \o{} to specify codepoints in octal, making it
   possible to specify values greater than 0777 and also making them
   unambiguous.

.  In UCP mode, \s was not matching two of the characters that Perl matches,
   namely NEL (U+0085) and MONGOLIAN VOWEL SEPARATOR (U+180E), though they
   were matched by \h.

.  Add JIT support for the 64 bit TileGX architecture.

.  Upgraded the handling of the POSIX classes [:graph:], [:print:], and
   [:punct:] when PCRE_UCP is set so as to include the same characters as Perl
   does in Unicode mode.

.  Perl no longer allows group names to start with digits, so I have made this
   change also in PCRE.

.  Added support for [[:<:]] and [[:>:]] as used in the BSD POSIX library to
   mean "start of word" and "end of word", respectively, as a transition aid.


Release 8.33 28-May-2013
--------------------------

A number of bugs are fixed, and some performance improvements have been made.
There are also some new features, of which these are the most important:

.  The behaviour of the backtracking verbs has been rationalized and
   documented in more detail.

.  JIT now supports callouts and all of the backtracking verbs.

.  Unicode validation has been updated in the light of Unicode Corrigendum #9,
   which points out that "non characters" are not "characters that may not
   appear in Unicode strings" but rather "characters that are reserved for
   internal use and have only local meaning".

.  (*LIMIT_MATCH=d) and (*LIMIT_RECURSION=d) have been added so that the
   creator of a pattern can specify lower (but not higher) limits for the
   matching process.

.  The PCRE_NEVER_UTF option is available to prevent pattern-writers from using
   the (*UTF) feature, as this could be a security issue.


Release 8.32 30-November-2012
-----------------------------

This release fixes a number of bugs, but also has some new features. These are
the highlights:

.  There is now support for 32-bit character strings and UTF-32. Like the
   16-bit support, this is done by compiling a separate 32-bit library.

.  \X now matches a Unicode extended grapheme cluster.

.  Case-independent matching of Unicode characters that have more than one
   "other case" now makes all three (or more) characters equivalent. This
   applies, for example, to Greek Sigma, which has two lowercase versions.

.  Unicode character properties are updated to Unicode 6.2.0.

.  The EBCDIC support, which had decayed, has had a spring clean.

.  A number of JIT optimizations have been added, which give faster JIT
   execution speed. In addition, a new direct interface to JIT execution is

storage/maria/aria_dump_log.c
/* Copyright (C) 2007 MySQL AB & Sanja Belkin

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; version 2 of the License.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335 USA */

#define VER "1.1"
#include "maria_def.h"
#include <my_getopt.h>
#include <welcome_copyright_notice.h>

extern void translog_example_table_init();
static const char *load_default_groups[]= { "aria_dump_log",0 };
static void get_options(int *argc,char * * *argv);
#ifndef DBUG_OFF
#if defined(_WIN32)
const char *default_dbug_option= "d:t:i:O,\\aria_dump_log.trace";
#else
const char *default_dbug_option= "d:t:i:o,/tmp/aria_dump_log.trace";
#endif
#endif
static ulonglong opt_offset;
static ulong opt_pages;
static const char *opt_file= NULL;
static File handler= -1;
static my_bool opt_unit= 0;
static struct my_option my_long_options[] =
{
#ifdef IMPLEMENTED
  {"body", 'b',
   "Print chunk body dump",
   (uchar **) &opt_body, (uchar **) &opt_body, 0,
   GET_BOOL, NO_ARG, 0, 0, 0, 0, 0, 0},
#endif
#ifndef DBUG_OFF
  {"debug", '#', "Output debug log. Often the argument is 'd:t:o,filename'.",
   0, 0, 0, GET_STR, OPT_ARG, 0, 0, 0, 0, 0, 0},
#endif
  {"file", 'f', "Path to file which will be read",
    (uchar**) &opt_file, 0, 0, GET_STR, REQUIRED_ARG, 0, 0, 0, 0, 0, 0},
  {"help", '?', "Display this help and exit.",
   0, 0, 0, GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0},
  { "offset", 'o', "Start reading log from this offset",
    (uchar**) &opt_offset, (uchar**) &opt_offset,
    0, GET_ULL, REQUIRED_ARG, 0, 0, ~(longlong) 0, 0, 0, 0 },
  { "pages", 'n', "Number of pages to read",
    (uchar**) &opt_pages, (uchar**) &opt_pages, 0,
    GET_ULONG, REQUIRED_ARG, (long) ~(ulong) 0,
    (long) 1, (long) ~(ulong) 0, 0, (long) 1, 0},
  {"unit-test", 'U',
   "Use unit test record table (for logs created by unittests",
   (uchar **) &opt_unit, (uchar **) &opt_unit, 0,
   GET_BOOL, NO_ARG, 0, 0, 0, 0, 0, 0},
  {"version", 'V', "Print version and exit.",
   0, 0, 0, GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0},
  { 0, 0, 0, 0, 0, 0, GET_NO_ARG, NO_ARG, 0, 0, 0, 0, 0, 0}
};


static void usage(void)
{
  print_version();
  puts("Copyright (C) 2008 MySQL AB");
  puts("This software comes with ABSOLUTELY NO WARRANTY. This is free software,");
  puts("and you are welcome to modify and redistribute it under the GPL license\n");

  puts("Dump the raw content of aria log pages.");
  puts("For a logical dump, use aria_read_log");
  printf("\nUsage: %s OPTIONS aria_log_file\n", my_progname_short);
  my_print_help(my_long_options);
  print_defaults("my", load_default_groups);
  my_print_variables(my_long_options);
}


static my_bool
get_one_option(const struct my_option *opt,
               const char *argument __attribute__((unused)),
               const char *filename __attribute__((unused)))
{
  switch (opt->id) {
  case '?':
    usage();
    exit(0);
  case 'V':
    print_version();
    exit(0);
#ifndef DBUG_OFF
  case '#':
    DBUG_SET_INITIAL(argument ? argument : default_dbug_option);
    break;
#endif
  }
  return 0;
}


static void get_options(int *argc,char ***argv)
{
  int ho_error;

  if ((ho_error=handle_options(argc, argv, my_long_options, get_one_option)))
    exit(ho_error);

  if (opt_file == NULL && *argc == 1)
    opt_file= **argv;

  if (opt_file == NULL)
  {
    usage();
    exit(1);
  }
}


/**
  @brief maria_dump_log main function.
*/

int main(int argc, char **argv)
{
  char **default_argv;
  uchar buffer[TRANSLOG_PAGE_SIZE];
  MY_INIT(argv[0]);

  load_defaults_or_exit("my", load_default_groups, &argc, &argv);
  default_argv= argv;
  get_options(&argc, &argv);

  if (opt_unit)
    translog_example_table_init();
  else
    translog_table_init();
  translog_fill_overhead_table();

  maria_data_root= ".";

  if ((handler= my_open(opt_file, O_RDONLY, MYF(MY_WME))) < 0)
  {
    fprintf(stderr, "Can't open file: '%s'  errno: %d\n",
            opt_file, my_errno);
    goto err;
  }
  if (my_seek(handler, opt_offset, SEEK_SET, MYF(MY_WME)) !=
      opt_offset)
  {
     fprintf(stderr, "Can't set position %lld  file: '%s'  errno: %d\n",
             opt_offset, opt_file, my_errno);
     goto err;
  }
  for (;
       opt_pages;
       opt_offset+= TRANSLOG_PAGE_SIZE, opt_pages--)
  {
    if (my_pread(handler, buffer, TRANSLOG_PAGE_SIZE, opt_offset,
                 MYF(MY_NABP)))
    {
      if (my_errno == HA_ERR_FILE_TOO_SHORT)
        goto end;
      fprintf(stderr, "Can't read page at position %lld  file: '%s'  "
              "errno: %d\n", opt_offset, opt_file, my_errno);
      goto err;
    }
    printf("Page by offset %llu (0x%llx)\n", opt_offset, opt_offset);
    dump_page(buffer, handler);
  }

end:
  my_close(handler, MYF(0));
  free_defaults(default_argv);
  exit(0);
  return 0;				/* No compiler warning */

err:
  my_close(handler, MYF(0));
  fprintf(stderr, "%s: FAILED\n", my_progname_short);
  free_defaults(default_argv);
  exit(1);
}

#include "ma_check_standalone.h"

